File: spike/correlation-fastapi/src/routemap.py
"""
LAYER 1 — deterministic route dump (FastAPI).

FastAPI keeps the actual handler function object on every route, so the
symbol is directly available — no metadata reflection needed, unlike Nest.
This is the strongest form of layer 1: not a parse, not a guess, the live
routing table of the running application.
"""

import re
from dataclasses import dataclass
from typing import Optional

from fastapi.routing import APIRoute


@dataclass
class RouteMapping:
    method: str
    path_template: str          # '/api/invoices/{invoice_id}'
    framework: str              # 'fastapi'
    handler_symbol: str         # 'fixture.app.find_one'
    handler_module: str
    handler_function: str
    source: str                 # 'runtime_dump'


def dump_routes(app) -> list[RouteMapping]:
    routes: list[RouteMapping] = []

    for route in app.routes:
        if not isinstance(route, APIRoute):
            continue                      # skips /openapi.json, /docs, etc.

        endpoint = route.endpoint
        for method in sorted(route.methods - {"HEAD", "OPTIONS"}):
            routes.append(
                RouteMapping(
                    method=method,
                    # FastAPI already stores brace form; Nest stores ':id'.
                    # Both sides of the join must normalize identically.
                    path_template=route.path,
                    framework="fastapi",
                    handler_symbol=f"{endpoint.__module__}.{endpoint.__qualname__}",
                    handler_module=endpoint.__module__,
                    handler_function=endpoint.__qualname__,
                    source="runtime_dump",
                )
            )

    return sorted(routes, key=lambda r: r.path_template)


def templatize(observed_path: str, routes: list[RouteMapping]) -> Optional[str]:
    """
    Maps a concrete observed path back to its template.

    Note the parameter NAME differs across frameworks for the same endpoint
    ('{id}' in the Nest fixture, '{invoice_id}' here). Matching must therefore
    be name-agnostic — the template is canonicalized by the route, never by
    string equality against a crawler's output.
    """
    for route in routes:
        pattern = "^" + re.sub(r"\{[A-Za-z0-9_]+\}", "[^/]+", route.path_template) + "$"
        if re.match(pattern, observed_path):
            return route.path_template
    return None

File: spike/correlation/src/routemap.ts
import { INestApplication } from '@nestjs/common';
import { PATH_METADATA, METHOD_METADATA } from '@nestjs/common/constants';
import { RequestMethod } from '@nestjs/common';
import { RouteMapping } from './types';

/**
 * LAYER 1 — deterministic route dump.
 *
 * Reads Nest's own decorator metadata off the live application's DI
 * container. This is ground truth, not a static parse: these are the
 * controllers Nest actually instantiated and is actually routing to.
 *
 * Why this beats parsing Express's `_router.stack`: the Express layer
 * knows the path but only carries an anonymous bound function, so the
 * handler identity — `InvoiceController.findOne` — is lost. Correlation
 * needs the symbol, so we read it where it still exists.
 */
export function dumpRoutes(app: INestApplication): RouteMapping[] {
  const container = (app as any).container;
  const modules: Map<unknown, any> = container.getModules();
  const routes: RouteMapping[] = [];

  for (const [, moduleRef] of modules) {
    for (const [, wrapper] of moduleRef.controllers as Map<unknown, any>) {
      const instance = wrapper.instance;
      if (!instance) continue;

      const ctor = instance.constructor;
      const controllerPath: string = Reflect.getMetadata(PATH_METADATA, ctor) ?? '';

      const proto = Object.getPrototypeOf(instance);
      for (const name of Object.getOwnPropertyNames(proto)) {
        if (name === 'constructor') continue;
        const handler = proto[name];
        if (typeof handler !== 'function') continue;

        const methodPath = Reflect.getMetadata(PATH_METADATA, handler);
        if (methodPath === undefined) continue;   // not a route handler

        const verbEnum = Reflect.getMetadata(METHOD_METADATA, handler);
        const method = RequestMethod[verbEnum] ?? 'GET';

        routes.push({
          method: method === 'ALL' ? 'GET' : method,
          pathTemplate: normalizePath(controllerPath, methodPath),
          framework: 'nestjs',
          handlerSymbol: `${ctor.name}.${name}`,
          controllerClass: ctor.name,
          handlerMethod: name,
          source: 'runtime_dump',
        });
      }
    }
  }

  return routes.sort((a, b) => a.pathTemplate.localeCompare(b.pathTemplate));
}

/**
 * Nest writes ':id'. The endpoint inventory keys on '{id}'. Both sides of
 * the join must normalize identically or nothing ever matches — see
 * docs/06-recon on path-template normalization.
 */
export function normalizePath(controllerPath: string, methodPath: string): string {
  const joined = `/${controllerPath}/${methodPath}`
    .replace(/\/+/g, '/')
    .replace(/\/$/, '');
  const withBraces = joined.replace(/:([A-Za-z0-9_]+)/g, '{$1}');
  return withBraces === '' ? '/' : withBraces;
}

/** Normalizes a concrete observed URL path to its template form. */
export function templatize(observedPath: string, routes: RouteMapping[]): string | null {
  for (const route of routes) {
    const pattern = new RegExp(
      '^' + route.pathTemplate.replace(/\{[A-Za-z0-9_]+\}/g, '[^/]+') + '$',
    );
    if (pattern.test(observedPath)) return route.pathTemplate;
  }
  return null;
}
